### PR TITLE
BUG: stats: Fix handling of float32 inputs for the boost-based ufuncs.

### LINKED ISSUE #15961: BUG: scipy.stats.beta and bernoulli fails with float32 inputs
### Describe your issue.

Some `scipy.stats` distribution, such as `beta` and `bernoulli` fail to compute statistical values (such as `median`, `var`, etc.) when the input is in a float32 format. The error does not occur with float64 (but we care about memory on our end).

The error varies if the float32 inputs are whole numbers (like `2.` or `3.`), which causes a NaN return, or decimal numbers (like `2.1` or `3.14`), which results in an tiny number of ~1e-20.

From debugging, our suspect is that the boost library is causing the problem.

The issue arises on an Intel Mac, with the code running in a docker container.

### Reproducing Code Example

```python
from scipy.stats import beta, bernoulli
import numpy as np

#create float64 input
a64 = np.array(2,dtype=float)
b64 = np.array(3,dtype=float)

#create float32 input
a32 = np.array(2,dtype=np.float32)
b32 = np.array(3,dtype=np.float32)

# those appears the same
assert a64 == a32

# defining the functions work
beta64 = beta(a64,b64)
beta32 = beta(a32,b32)

# getting statistical value for 64 works
print("beta true value: ", beta(a64,b64).mean())

# but fails for 32 (NaN outcome)
print("beta fails with float32, integer numbers inputs; outcome: ", beta(a32,b32).mean())

# using decimal numbers also change the outcome (irrealistic, wrong tiny number, with overflow warning)
a32_dec = np.array(2.1,dtype=np.float32)
b32_dec = np.array(3.14,dtype=np.float32)
print("beta fails with float32, decimal numbers inputs; outcome: ", beta(a32_dec,b32_dec).mean())

#demonstration with Bernoulli
p = 0.33
print("Bernoulli: true value is {}; outcome is {}".format(p, bernoulli(np.array(p,dtype=np.float32)).mean()))
```

### Error message

```shell
beta true value:  0.4
beta fails with float32, integer numbers inputs; outcome: nan 
beta fails with float32, decimal numbers inputs; outcome: -5.421010862427522e-20
Bernoulli: true value is 0.33; outcome is 0.0

### error for integer values as float32
/usr/local/lib/python3.9/site-packages/scipy/stats/_continuous_distns.py:628: RuntimeWarning: invalid value encountered in _beta_mean
  _boost._beta_mean(a, b),
/usr/local/lib/python3.9/site-packages/scipy/stats/_continuous_distns.py:629: RuntimeWarning: invalid value encountered in _beta_variance
  _boost._beta_variance(a, b),
/usr/local/lib/python3.9/site-packages/scipy/stats/_continuous_distns.py:630: RuntimeWarning: invalid value encountered in _beta_skewness
  _boost._beta_skewness(a, b),
/usr/local/lib/python3.9/site-packages/scipy/stats/_continuous_distns.py:631: RuntimeWarning: invalid value encountered in _beta_kurtosis_excess
  _boost._beta_kurtosis_excess(a, b))

### error for decimal float32 values
/usr/local/lib/python3.9/site-packages/scipy/stats/_continuous_distns.py:629: RuntimeWarning: overflow encountered in _beta_variance
  _boost._beta_variance(a, b),
/usr/local/lib/python3.9/site-packages/scipy/stats/_continuous_distns.py:631: RuntimeWarning: overflow encountered in _beta_kurtosis_excess
  _boost._beta_kurtosis_excess(a, b))
```

### SciPy/NumPy/Python version information

1.7.1 1.21.5 sys.version_info(major=3, minor=9, micro=10, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
The C++14 code for calling the underlying C++ function that does
the actual scalar calculation for a ufunc had a bug where the
float arguments were implicitly cast to double while the function
expected float, resulting in incorrect output.

The fix here is to remove the C++14 version of the code, and
use only the less elegant C++11 code.

Closes gh-15961

### CODE REVIEW

**Comment by WarrenWeckesser** [comment]
With just the first commit of this PR, the only tests that failed were the new boost ufunc tests for the data type `np.longdouble` on the Windows Meson platform.  The log showed that the tests start failing when the data type of the inputs was `np.longdouble`, and after a few failures, Python crashed with an access violation.  I suspect this is related to how the `long double` type is handled by the compiler.  It appears that on the Meson Windows platform, gcc is used to build SciPy, and C files (but not C++ files) are compiled with the option `-mlong-double-64` (for compatibility with the MSVC compiler):

https://github.com/scipy/scipy/blob/084df9925bb95cffc5429a5da997b6501d17d249/scipy/meson.build#L6-L13

We now have C++ code that works with the `long double` type, so it seems we should compile *all* C and  C++ code with the same option (at least all code that uses the `long double` type). I suspect something is not right in how this option is being used currently, but at the moment I don't know what the correct settings should be.

In a follow-up commit, I tried changing the global compiler arguments with
```
    add_global_arguments('-mlong-double-64', language: ['c', 'cpp'])
```
With that change, the test suite crashed at the first import of the boost ufunc extension module in stats, before any actual tests were run.

I then tried tweaking `meson.build` in `scipy/stats/_boost` (the latest commit has one of my shots in the dark), but still got a crash on the import of the boost ufunc extension modules.

In summary: the first commit here fixes gh-15961, but the fix seems to expose a problem with how we handle the `long double` type on the Windows platform when building with mingw.

**Comment by WarrenWeckesser** [comment]
In my last commit, I fixed the problem with the `long double` data type on the Windows Meson job (which uses mingw) by not generating `long double` ufunc loops in Windows for the `stats` boost-based ufuncs.  So on Windows, a ufunc such as `scipy.stats._boost._beta_cdf` will have the ufunc type signatures `['fff->f', 'ddd->d']`; on other platforms, the signatures will also include `'ggg->g'`.  Since on the MSVC builds, `long double` is the same as the `double`, there is no real loss by doing this.  The only failure now is the one unrelated timeout.

**Comment by rgommers** [comment]
@WarrenWeckesser it seems like the plan was to remove the C++11 code instead (Cc @mckib2 who wrote this code). I believe we can now use C++14 unconditionally, see http://scipy.github.io/devdocs/dev/toolchain.html#c-language-standards.

**Comment by rgommers** [comment]
> In my last commit, I fixed the problem with the `long double` data type on the Windows Meson job (which uses mingw) by not generating `long double` ufunc loops in Windows for the `stats` boost-based ufuncs. So on Windows, a ufunc such as `scipy.stats._boost._beta_cdf` will have the ufunc type signatures `['fff->f', 'ddd->d']`; on other platforms, the signatures will also include `'ggg->g'`. Since on the MSVC builds, `long double` is the same as the `double`, there is no real loss by doing this. The only failure now is the one unrelated timeout.

This seems quite reasonable to me. Over time I'd expect we will drop `long double` completely, since it's more or less obsolete (not only on Windows).

**Comment by rgommers** [comment]
> @WarrenWeckesser it seems like the plan was to remove the C++11 code instead (Cc @mckib2 who wrote this code). I believe we can now use C++14 unconditionally,

Ah, I just saw your detailed comment about trying to fix the C++14 version and ending up with very similar code to the current C++11 implementation. So this is fine - and if needed/desired, the C++14 version can be reinstated.

I completely agree that we should not have two code paths here depending on compile options / compiler capability.

**Comment by WarrenWeckesser** [comment]
Right, the details are explained in https://github.com/scipy/scipy/issues/15961#issuecomment-1126786894.  The C++11 code is also compatible with C++14, so it is perhaps better to just call it the working C++ code.  To be clear, accepting this PR doesn't mean the code can't be compiled with C++14.

**Comment by rgommers** [comment]
> In a follow-up commit, I tried changing the global compiler arguments with
> 
> ```
>     add_global_arguments('-mlong-double-64', language: ['c', 'cpp'])
> ```
> 
> With that change, the test suite crashed at the first import of the boost ufunc extension module in stats, before any actual tests were run.
> 
> I then tried tweaking `meson.build` in `scipy/stats/_boost` (the latest commit has one of my shots in the dark), but still got a crash on the import of the boost ufunc extension modules.
> 
> In summary: the first commit here fixes [gh-15961](https://github.com/scipy/scipy/issues/15961), but the fix seems to expose a problem with how we handle the `long double` type on the Windows platform when building with mingw.

It's not clear to me from the code comments in `scipy/meson.build` why we cannot add `-mlong-double-64` for C++, or why it isn't needed. It'd be good to extend that comment a bit so that is clearer to the next person who runs into a C++ longdouble issue - @carlkl, matthew-brett can you comment on this?

**Review by rgommers — APPROVED** [review]
This LGTM and fixes a correctness issue, so let's get it in for 1.9.0. Thanks @WarrenWeckesser!

**Comment by mckib2** [comment]
> > @WarrenWeckesser it seems like the plan was to remove the C++11 code instead (Cc @mckib2 who wrote this code). I believe we can now use C++14 unconditionally,
> 
> Ah, I just saw your detailed comment about trying to fix the C++14 version and ending up with very similar code to the current C++11 implementation. So this is fine - and if needed/desired, the C++14 version can be reinstated.
> 
> I completely agree that we should not have two code paths here depending on compile options / compiler capability.

Sorry, just getting back from long work travel and sickness.  The C++14 solution always seemed "cleaner" to me with the auto deduction template pack capabilities, but if it's faulty or doesn't work right, I agree removing it for the C++11 was the way to go here.  It may however need to be expanded in the future if there are any distributions that require more arguments than are currently supported
